### DISCUSSION TITLE
Is it possible to use filed value from one schema as a field name in another schema?

### DISCUSSION THREAD

**Discussion by amaslakov** [Automattic/mongoose#9979] [Q&A] [answered]
Eventually I want to fill field names in one collection with field values from another collection. Like I have a collection of some items:
```
{ _id: ..., name: "item1", parameter1: "foo", parameter2: "bar", ... etc.},
{ _id: ..., name: "item2", parameter1: "42", paraneter2: "1337", ... etc.},
```
I want another collection to have fields with item names from first collection:
```
{ _id: ..., "item1": { amount: 42, locked: false }, "item2": { amount: 1337, locked: true }, ... etc.}
```
How can I do it Schema-wise for the second collection? Thanks

**Comment by vkarpov15** [comment, marked as answer]
It looks like you're looking for [maps](https://mongoosejs.com/docs/schematypes.html#maps)

**Reply by amaslakov** [reply]
Thanks for your answer! But I don't quite understand, how can a map enforce the connection between field value from collection1 and field name in collection2?

**Reply by vkarpov15** [reply]
I don't understand what you mean, can you please elaborate?

**Reply by amaslakov** [reply]
Let's say I have a list of properties (represented in collection 1):
```
{ 'name': 'humidity', 'units': 'percent', 'isRegulatedByStandard1': true },
{ 'name': 'fat', 'units': 'percent', 'isRegulatedByStandard1': true },
{ 'name': 'protein', 'units': 'percent', 'isRegulatedByStandard1': true },
```
And then I have a collection of ingredients, that all have to reference that collection of properties:
```
{ 'name': 'powder', 'humidity': 5, 'fat': 0, 'protein': 12 },
{ 'name': 'eggs', 'humidity': 70, 'fat': 4, 'protein': 15 },
{ 'name': 'milk', 'humidity': 90, 'fat': 6, 'protein': 4 },
```
First I thought about reference by name (since property names are unique), but it may be better to reference those by ID. I made up all the numbers, so don't look to much at those =)
Anyways, the idea is that all ingredients have the same set of properties, but different amounts of them. I want to be able to edit an ingredient's properties from the front end and I wanted mongoose to do the validation for me (don't pass properties if those aren't described by schema). I want that schema to update if I add a new property to the properties collection. Otherwise, I would have to do the validation manually by reading that properties table. I wonder if it is possible to do that with mongoose.
Now when I wrote it, it sounds like I want too much =)

**Reply by vkarpov15** [reply]
There isn't a good way to update a model's schema once you've created a model, so your best bet would be to use a Map and write a custom validator that loads from the `Property` model. Below is some pseudocode that demonstrates the idea:

```javascript
properties: {
  type: Map,
  of: number,
  required: true,
  validate: async function(val) {
    const propertiesInDb = mongoose.model('Property').find();
    const propertiesByName = new Set(propertiesInDb.map(p => p.name));
    for (const key of val.keys()) {
      if (!propertiesByName.has(key)) {
        throw new Error(`Key ${key} is not a valid ingredient property`);
      }
    }
  }
}
```

**Reply by amaslakov** [reply]
Okay, thank you!
